src/formatting/items.rs

```diff
@@ -1131,30 +1131,42 @@ pub(crate) fn format_trait(
         );
         result.push_str(&header);
 
+        // FIXME: rustfmt fails to format when there are comments before the ident.
+        if contains_comment(context.snippet(mk_sp(item.span.lo(), generics.span.lo()))) {
+            return None;
+        }
+
         let body_lo = context.snippet_provider.span_after(item.span, "{");
 
         let shape = Shape::indented(offset, context.config).offset_left(result.len())?;
         let generics_str =
             rewrite_generics(context, rewrite_ident(context, item.ident), generics, shape)?;
         result.push_str(&generics_str);
 
-        // FIXME(#2055): rustfmt fails to format when there are comments between trait bounds.
+        // FIXME(#2055): rustfmt fails to format when there are comments within trait bounds.
         if !generic_bounds.is_empty() {
-            let ident_hi = context
-                .snippet_provider
-                .span_after(item.span, &item.ident.as_str());
+            let bound_lo = generic_bounds.first().unwrap().span().lo();
             let bound_hi = generic_bounds.last().unwrap().span().hi();
-            let snippet = context.snippet(mk_sp(ident_hi, bound_hi));
+            let snippet = context.snippet(mk_sp(bound_lo, bound_hi));
             if contains_comment(snippet) {
                 return None;
             }
```

Comment:
I'm not sure if this is true/needed anymore given #4474 (and the fix/adjustment in #4653). I think that currently (with the inclusion of the correction to the pre-block span derivation) we can already handle any and all comments to the right of the colon following the ident.

I think the remaining gaps are really any such comments to the left

Reply:
It seems that you are correct and comments whiting that generics bounds are handled properly.   Before I make the change, there is one change to **2015 edition** test case output that I want to make sure is acceptable.  The following code: https://github.com/rust-lang/rustfmt/blob/0e908555273391bc3aba087abae72791451fd84a/tests/source/trait_2015_edition.rs#L65-L70
Will now be formatted as (same as for the other editions):
```rust
pub trait Foo:
    // A and C
    A
        + C
        // and B
        + B
{
}
```
**Is this o.k.** or the test for comments within the generics bound should still be done when edition is set to 2015?

Reply:
The edition really only drives parsing, and shouldn't have any impact on formatting. I wonder if there's a separate bug in there :thinking: 

What happens if you remove that leading `// A and C` comment?

Reply:
> What happens if you remove that leading `// A and C comment`?

You are right that removing the `// A and C comment` changes the indentation to:
```rustfmt
pub trait Foo:
    A
    + C
    // and B
    + B
{
}
```
> I wonder if there's a separate bug in there.

There seem to be a general a problem in `rewrite_assign_rhs_with_comments()`.  The same issue is seen in the following assignments, where an extra indentation is added because of a comment the the 2nd and following lines of the expression (the long names are because adding the `// and B` comment into the expressions cause using the original code):
```rust
fn main () {
	x =
	Aaaaaaaaaaaaaaaa   +    Cccccccccccccccccccccccccccc
          +    Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb ;
	y =     /* A and C */
	Aaaaaaaaaaaaaaaa   +    Cccccccccccccccccccccccccccc
          +    Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb ;
	z =     
	/* A and C */
	Aaaaaaaaaaaaaaaa   +    Cccccccccccccccccccccccccccc
          +    Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb ;
}

```
is formatted as:
```rust
fn main() {
    x = Aaaaaaaaaaaaaaaa
        + Cccccccccccccccccccccccccccc
        + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
    y = /* A and C */
        Aaaaaaaaaaaaaaaa
            + Cccccccccccccccccccccccccccc
            + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
    z =
        /* A and C */
        Aaaaaaaaaaaaaaaa
            + Cccccccccccccccccccccccccccc
            + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
}
```
The problem is caused by the following `shape` adjustment that should probably be done only if the formatted comment is not immediately followed by newline: https://github.com/rust-lang/rustfmt/blob/e10d2d77180a6030d4dfa0d8064357cf2ca0c266/src/formatting/expr.rs#L1998-L2002

**Should I try to fix this issue in this PR, or for now the test case expected target should just be changed?** (Note that the test case output will be changed in any case, since now it is using the original code and with this PR the code is formatted.)

Reply:
> The same issue is seen in the following assignments, where an extra indentation is added because of a comment the the 2nd and following lines of the expression

Well, not quite. The former are bounds on a trait whereas your example are binop expressions, and these are governed by different formatting rules.

https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/items.md#traits vs https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/expressions.md#binary-operations

As such the only issue/non-compliance with the Style Guide is on the traits, so I think you'll want to try to address that in the same PR since both that fix and the originally discussed changes here would be needed in combination (neither would be mergeable in isolation)

Reply:
> The former are bounds on a trait whereas your example are binop expressions, and these are governed by different formatting rules.

I don't see a reference in the guide to the formatting output I was referring to, so I will try to clarify what I meant.  I expected that the `y` assignment in the example above will be similar to the `x` assignment and to the formatting example in the guide:
```rust
    y = /* A and C */
        Aaaaaaaaaaaaaaaa
        + Cccccccccccccccccccccccccccc
        + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
```
However, because of the comment the 3rd and 4th lines are formatted with additional indentation:
```rust
    y = /* A and C */
        Aaaaaaaaaaaaaaaa
            + Cccccccccccccccccccccccccccc
            + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
```
Is this the expected output?

Reply:
> I expected that the y assignment in the example above will be similar to the x assignment and to the formatting example in the guide

What I'm saying is that you've changed to the (off)topic discussion of binary expressions which is not relevant for the formatting of trait bounds, and additionally that your expectation about binary expression formatting is incorrect.

From the binary expression section of the guide (emphasis mine):

> If line-breaking, put the operator on a new line **and block indent**. Put each sub-expression on its own line

The block indents in the binary expressions are happening because they are _supposed_ to be there, the only impact the comment between the assignment operator and the rhs is the placement of the first operand. If another condition (such as a comment) results in the opening of the binary expression not being able to be placed on the same line as the lhs/assignment operator, then that first operand of the bin expr has to be block indented itself, but the other subexpr(s) still have to be block indented relative to that first operand.

It is merely a coincidence in your first snippet that the first operand happens to be in the same column position as the block indentation, due to the single character variable name (`x` + `=` + two spaces). If you were to use a longer var name, have a `let`, etc. you will see that the alignment is different:

e.g. try formatting these:
```rust
fn main() {
    let x = Aaaaaaaaaaaaaaaa
        + Cccccccccccccccccccccccccccc
        + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
        
        foo_bar = Aaaaaaaaaaaaaaaa
        + Cccccccccccccccccccccccccccc
        + Bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
}
```

Hopefully that helps, but if you have additional thoughts/questions/doubts about the binary expression formatting, then I'd encourage opening a separate issue to avoid taking this thread any farther away from the issue at hand, namely the extra, incorrect additional block indent that's added to the trait bounds when there's a comment preceding the first bound.

Reply:
Submitted changes which I hope satisfy the comments requirements.

> Hopefully that helps, but if you have additional thoughts/questions/doubts about the binary expression formatting, then I'd encourage opening a separate issue

Yes, I understand now.  Thanks for the detailed explanation.   I do have some minor comments, but its not worth further discussion, especially considering backward compatibility.

> What I'm saying is that you've changed to the (off)topic discussion of binary expressions which is not relevant for the formatting of trait bounds

I know it was (of)topic but I thought that a common change for both traits and binop is required.  I now added a separate function to handle trait rhs with comments.